File: adk-go/internal/airline/repository/rebooking.go
package repository

import (
	"context"
	"errors"
	"fmt"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
	airline "github.com/tokiou/agents-at-scale/internal/airline"
	"github.com/tokiou/agents-at-scale/internal/platform/postgres/sqlc"
)

type RebookingRepository struct {
	queries *sqlc.Queries
	pool    *pgxpool.Pool
}

func NewRebookingRepository(queries *sqlc.Queries, pool *pgxpool.Pool) *RebookingRepository {
	return &RebookingRepository{queries: queries, pool: pool}
}

func (r *RebookingRepository) GetSnapshot(ctx context.Context, selection airline.RebookingSelection) (*airline.RebookingSnapshot, *airline.TravelCredit, error) {
	return r.snapshot(ctx, r.queries, selection)
}

func (r *RebookingRepository) snapshot(ctx context.Context, queries *sqlc.Queries, selection airline.RebookingSelection) (*airline.RebookingSnapshot, *airline.TravelCredit, error) {
	segment, err := queries.GetRebookingSegmentForUpdate(ctx, selection.SegmentID)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil, ErrSegmentNotFound
	}
	if err != nil {
		return nil, nil, fmt.Errorf("get rebooking segment: %w", err)
	}
	target, err := queries.GetRebookingTargetForUpdate(ctx, sqlc.GetRebookingTargetForUpdateParams{ID: selection.NewFlightID, FareClassID: selection.NewFareClassID})
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil, ErrFlightNotFound
	}
	if err != nil {
		return nil, nil, fmt.Errorf("get rebooking target: %w", err)
	}
	snapshot := &airline.RebookingSnapshot{
		SegmentID: segment.SegmentID, ReservationID: segment.ReservationID, CustomerID: segment.CustomerID,
		OldFlightID: segment.OldFlightID, OldFareClassID: segment.OldFareClassID,
		SegmentStatus: airline.SegmentStatus(segment.SegmentStatus), ReservationStatus: airline.ReservationStatus(segment.ReservationStatus),
		OldFlightStatus: airline.FlightStatus(segment.OldFlightStatus), OriginAirport: segment.OriginAirport,
		DestinationAirport: segment.DestinationAirport, OldDepartureAt: segment.OldDepartureAt.Time,
		OldPrice: segment.PricePaid, OldCurrency: segment.SegmentCurrency, OldChangeAllowed: segment.OldChangeAllowed,
		PassengerCount: int(segment.PassengerCount),
		TargetFlightID: target.FlightID, TargetFareClassID: target.FareClassID, TargetStatus: airline.FlightStatus(target.FlightStatus),
		TargetOrigin: target.OriginAirport, TargetDestination: target.DestinationAirport, TargetDepartureAt: target.DepartureAt.Time,
		TargetPrice: target.Price, TargetCurrency: target.Currency, AvailableSeats: int(target.AvailableSeats),
		TargetChangeAllowed: target.ChangeAllowed, TargetChangeFee: target.ChangeFee,
	}
	if selection.TravelCreditID == uuid.Nil {
		return snapshot, nil, nil
	}
	credit, err := queries.GetTravelCreditForUpdate(ctx, sqlc.GetTravelCreditForUpdateParams{ID: selection.TravelCreditID, CustomerID: segment.CustomerID})
	if errors.Is(err, pgx.ErrNoRows) {
		return snapshot, nil, ErrTravelCreditNotFound
	}
	if err != nil {
		return nil, nil, fmt.Errorf("get travel credit: %w", err)
	}
	value := mapTravelCredit(credit)
	return snapshot, &value, nil
}

func (r *RebookingRepository) Execute(ctx context.Context, selection airline.RebookingSelection) (*airline.RebookingResult, error) {
	tx, err := r.pool.Begin(ctx)
	if err != nil {
		return nil, fmt.Errorf("begin rebooking transaction: %w", err)
	}
	defer tx.Rollback(ctx)
	queries := r.queries.WithTx(tx)
	snapshot, credit, err := r.snapshot(ctx, queries, selection)
	if err != nil {
		return nil, err
	}
	validation, err := airline.ValidateRebookingSelection(selection, *snapshot, credit)
	if err != nil {
		return nil, fmt.Errorf("validate rebooking selection: %w", err)
	}
	if err := queries.DecrementFlightFareSeats(ctx, sqlc.DecrementFlightFareSeatsParams{PassengerCount: int32(validation.PassengerCount), FlightID: validation.NewFlightID, FareClassID: validation.NewFareClassID}); err != nil {
		return nil, fmt.Errorf("decrement rebooking inventory: %w", err)
	}
	if err := queries.IncrementFlightFareSeats(ctx, sqlc.IncrementFlightFareSeatsParams{PassengerCount: int32(validation.PassengerCount), FlightID: validation.OldFlightID, FareClassID: validation.OldFareClassID}); err != nil {
		return nil, fmt.Errorf("restore original inventory: %w", err)
	}
	if err := queries.UpdateReservationSegment(ctx, sqlc.UpdateReservationSegmentParams{ID: selection.SegmentID, FlightID: validation.NewFlightID, FareClassID: validation.NewFareClassID, PricePaid: validation.NewPrice, Currency: validation.Currency}); err != nil {
		return nil, fmt.Errorf("update reservation segment: %w", err)
	}
	if err := queries.UpdateReservationTotal(ctx, sqlc.UpdateReservationTotalParams{ID: validation.ReservationID, TotalAmount: validation.FareDifference.Add(validation.ChangeFee).Sub(validation.TravelCreditUsed)}); err != nil {
		return nil, fmt.Errorf("update reservation total: %w", err)
	}
	if selection.TravelCreditID != uuid.Nil {
		if err := queries.ConsumeTravelCredit(ctx, sqlc.ConsumeTravelCreditParams{ID: selection.TravelCreditID, RemainingAmount: selection.TravelCreditAmount}); err != nil {
			return nil, fmt.Errorf("consume travel credit: %w", err)
		}
	}
	change, err := queries.CreateFlightChange(ctx, sqlc.CreateFlightChangeParams{ID: uuid.New(), ReservationSegmentID: selection.SegmentID, OldFlightID: validation.OldFlightID, NewFlightID: validation.NewFlightID, OldFareClassID: validation.OldFareClassID, NewFareClassID: validation.NewFareClassID, FareDifference: validation.FareDifference, ChangeFee: validation.ChangeFee, TravelCreditUsed: validation.TravelCreditUsed, Currency: validation.Currency})
	if err != nil {
		return nil, fmt.Errorf("record flight change: %w", err)
	}
	if err := tx.Commit(ctx); err != nil {
		return nil, fmt.Errorf("commit rebooking transaction: %w", err)
	}
	return &airline.RebookingResult{Change: mapFlightChange(change), Segment: airline.ReservationSegment{ID: selection.SegmentID, ReservationID: validation.ReservationID, FlightID: validation.NewFlightID, FareClassID: validation.NewFareClassID, Status: airline.SegmentStatusChanged, PricePaid: validation.NewPrice, Currency: validation.Currency}}, nil
}

func (r *RebookingRepository) Verify(ctx context.Context, selection airline.RebookingSelection) (*airline.RebookingResult, error) {
	row, err := r.queries.GetRebookingResult(ctx, selection.SegmentID)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, ErrSegmentNotFound
	}
	if err != nil {
		return nil, fmt.Errorf("verify rebooking: %w", err)
	}
	if row.FlightID != selection.NewFlightID || row.FareClassID != selection.NewFareClassID || row.SegmentStatus != sqlc.SegmentStatusCHANGED {
		return nil, airline.ErrRebookingVerification
	}
	return &airline.RebookingResult{BookingReference: row.BookingReference, Segment: airline.ReservationSegment{ID: row.SegmentID, ReservationID: row.ReservationID, FlightID: row.FlightID, FareClassID: row.FareClassID, Status: airline.SegmentStatus(row.SegmentStatus), PricePaid: row.PricePaid, Currency: row.Currency}}, nil
}

File: langgraph-python/app/airline/repository/rebooking.py
"""Transactional repository for flight rebooking."""

from collections.abc import Callable
from datetime import UTC, datetime
from decimal import Decimal
from uuid import UUID, uuid4

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker
from sqlalchemy.orm import joinedload

from app.airline.models import (
    FlightFare,
    FlightStatus,
    FlightChange,
    Reservation,
    ReservationSegment,
    SegmentStatus,
    TravelCredit,
    TravelCreditStatus,
)
from app.airline.repository.errors import (
    FlightNotFoundError,
    RebookingVerificationError,
    SegmentNotFoundError,
    TravelCreditNotFoundError,
)
from app.airline.schemas import (
    FlightChangeSchema,
    RebookingResultSchema,
    RebookingSelectionSchema,
    RebookingSnapshotSchema,
    RebookingValidationSchema,
    ReservationSegmentSchema,
    TravelCreditSchema,
)


class RebookingRepository:
    def __init__(self, sessions: async_sessionmaker[AsyncSession]) -> None:
        self._sessions = sessions

    async def get_snapshot(
        self, selection: RebookingSelectionSchema
    ) -> tuple[RebookingSnapshotSchema, TravelCreditSchema | None]:
        async with self._sessions() as session:
            snapshot, credit = await self._snapshot(session, selection, lock=False)
            return snapshot, credit

    async def execute(
        self,
        selection: RebookingSelectionSchema,
        validate: Callable[[RebookingSelectionSchema, RebookingSnapshotSchema, TravelCreditSchema | None], RebookingValidationSchema],
    ) -> RebookingResultSchema:
        async with self._sessions() as session:
            async with session.begin():
                snapshot, credit = await self._snapshot(session, selection, lock=True)
                await self._lock_fares(session, snapshot.old_flight_id, snapshot.old_fare_class_id, selection)
                snapshot, credit = await self._snapshot(session, selection, lock=True)
                validation = validate(selection, snapshot, credit)
                target = await self._target(session, selection, lock=True)
                if target.available_seats < validation.passenger_count:
                    raise ValueError("target flight does not have enough seats")

                target.available_seats -= validation.passenger_count
                old_fare = await session.scalar(
                    select(FlightFare)
                    .where(
                        FlightFare.flight_id == validation.old_flight_id,
                        FlightFare.fare_class_id == validation.old_fare_class_id,
                    )
                    .with_for_update()
                )
                if old_fare is None:
                    raise FlightNotFoundError("original fare not found")
                old_fare.available_seats += validation.passenger_count

                segment = await session.get(ReservationSegment, selection.segment_id)
                reservation = await session.get(Reservation, validation.reservation_id)
                if segment is None or reservation is None:
                    raise SegmentNotFoundError("reservation segment not found")
                segment.flight_id = validation.new_flight_id
                segment.fare_class_id = validation.new_fare_class_id
                segment.price_paid = validation.new_price
                segment.currency = validation.currency
                segment.status = SegmentStatus.CHANGED
                segment.updated_at = datetime.now(UTC)
                reservation.total_amount += validation.fare_difference + validation.change_fee - validation.travel_credit_used
                reservation.updated_at = datetime.now(UTC)

                credit_entity = None
                if selection.travel_credit_id is not None:
                    credit_entity = await session.get(TravelCredit, selection.travel_credit_id, with_for_update=True)
                    if credit_entity is None:
                        raise TravelCreditNotFoundError("travel credit not found")
                    credit_entity.remaining_amount -= selection.travel_credit_amount
                    credit_entity.status = (
                        TravelCreditStatus.USED
                        if credit_entity.remaining_amount == 0
                        else TravelCreditStatus.PARTIALLY_USED
                    )

                change = FlightChange(
                    id=uuid4(),
                    reservation_segment_id=selection.segment_id,
                    old_flight_id=validation.old_flight_id,
                    new_flight_id=validation.new_flight_id,
                    old_fare_class_id=validation.old_fare_class_id,
                    new_fare_class_id=validation.new_fare_class_id,
                    fare_difference=validation.fare_difference,
                    change_fee=validation.change_fee,
                    travel_credit_used=validation.travel_credit_used,
                    currency=validation.currency,
                    created_at=datetime.now(UTC),
                )
                session.add(change)
                await session.flush()
                return RebookingResultSchema(
                    change=FlightChangeSchema.model_validate(change),
                    segment=ReservationSegmentSchema.model_validate(segment),
                    booking_reference=reservation.booking_reference,
                )

    async def verify(self, selection: RebookingSelectionSchema) -> RebookingResultSchema:
        async with self._sessions() as session:
            segment = await session.get(ReservationSegment, selection.segment_id)
            if segment is None:
                raise SegmentNotFoundError("reservation segment not found")
            if (
                segment.flight_id != selection.new_flight_id
                or segment.fare_class_id != selection.new_fare_class_id
                or segment.status != SegmentStatus.CHANGED
            ):
                raise RebookingVerificationError("rebooking verification failed")
            reservation = await session.get(Reservation, segment.reservation_id)
            if reservation is None:
                raise RebookingVerificationError("rebooking reservation not found")
            return RebookingResultSchema(
                segment=ReservationSegmentSchema.model_validate(segment),
                booking_reference=reservation.booking_reference,
            )

    async def _target(self, session, selection, lock):
        statement = (
            select(FlightFare)
            .options(joinedload(FlightFare.flight), joinedload(FlightFare.fare_class))
            .where(
                FlightFare.flight_id == selection.new_flight_id,
                FlightFare.fare_class_id == selection.new_fare_class_id,
            )
        )
        if lock:
            statement = statement.with_for_update()
        target = await session.scalar(statement)
        if target is None:
            raise FlightNotFoundError("target flight or fare not found")
        return target

    async def _lock_fares(self, session, old_flight_id, old_fare_class_id, selection):
        keys = sorted(
            {
                (old_flight_id, old_fare_class_id),
                (selection.new_flight_id, selection.new_fare_class_id),
            },
            key=lambda item: (str(item[0]), str(item[1])),
        )
        for flight_id, fare_class_id in keys:
            await session.scalar(
                select(FlightFare)
                .where(FlightFare.flight_id == flight_id, FlightFare.fare_class_id == fare_class_id)
                .with_for_update()
            )

    async def _snapshot(self, session, selection, lock):
        statement = (
            select(ReservationSegment)
            .options(
                joinedload(ReservationSegment.reservation),
                joinedload(ReservationSegment.reservation).selectinload(Reservation.passengers),
                joinedload(ReservationSegment.flight),
                joinedload(ReservationSegment.fare_class),
            )
            .where(ReservationSegment.id == selection.segment_id)
        )
        if lock:
            statement = statement.with_for_update()
        segment = await session.scalar(statement)
        if segment is None:
            raise SegmentNotFoundError("reservation segment not found")
        target = await self._target(session, selection, False)
        passenger_count = len(segment.reservation.passengers)
        snapshot = RebookingSnapshotSchema(
            segment_id=segment.id,
            reservation_id=segment.reservation_id,
            customer_id=segment.reservation.customer_id,
            old_flight_id=segment.flight_id,
            old_fare_class_id=segment.fare_class_id,
            segment_status=segment.status,
            reservation_status=segment.reservation.status,
            old_flight_status=segment.flight.status,
            origin_airport=segment.flight.origin_airport,
            destination_airport=segment.flight.destination_airport,
            old_departure_at=segment.flight.departure_at,
            old_price=segment.price_paid,
            old_currency=segment.currency,
            old_change_allowed=segment.fare_class.change_allowed,
            target_flight_id=target.flight_id,
            target_fare_class_id=target.fare_class_id,
            target_status=target.flight.status,
            target_origin=target.flight.origin_airport,
            target_destination=target.flight.destination_airport,
            target_departure_at=target.flight.departure_at,
            target_price=target.price,
            target_currency=target.currency,
            available_seats=target.available_seats,
            target_change_allowed=target.fare_class.change_allowed,
            target_change_fee=target.fare_class.change_fee,
            passenger_count=passenger_count,
        )
        credit = None
        if selection.travel_credit_id is not None:
            credit_entity = await session.get(TravelCredit, selection.travel_credit_id, with_for_update=lock)
            if credit_entity is None:
                raise TravelCreditNotFoundError("travel credit not found")
            credit = TravelCreditSchema.model_validate(credit_entity)
        return snapshot, credit
